### DISCUSSION TITLE
MQTT in imperial units?

### DISCUSSION THREAD

**Discussion by chilman408** [teslamate-org/teslamate#230] [Ideas]
Is it possible to display the MQTT information in imperial units rather than C and km?

**Comment by adriankumpf** [comment]
Currently not. I'd like to have a topic for each of the different units in the future though. For example:

* `teslamate/cars/$car_id/inside_temp_C` and 
* `teslamate/cars/$car_id/inside_temp_F`

**Comment by chilman408** [comment]
> Currently not. I'd like to have a topic for each of the different units in the future though. For example:
> 
> * `teslamate/cars/$car_id/inside_temp_C` and
> * `teslamate/cars/$car_id/inside_temp_F`

Thanks!  Yes, supporting imperial units natively would be great (LOL, I know US is way behind).  For now, conversion within HomeAssistant itself work fine for me.

**Comment by gaLDWoRE** [comment]
What does MQTT stand for?

**Comment by chisale** [comment]
I'd like this as well. Any arithmetic operation in yaml/home-assistant is not really straightforward. Sending a unit specific topic seems cleaner solution.

**Comment by ymilord** [comment]
I'd like this as well. I am using MQTT /w Home Assistant and a ESP32. And both placed I have to do manual conversion. The option mention above supported both would be awesome.

**Comment by MarkRichardsNH** [comment]
I know some of you have worked this out, but for others that may not have, you can use a value template in home assistant to create an additional entity with the imperial units. 
Basically its reading the metric unit and doing the calculation. 

You can leave the original entries there and add these as extras (one example of each below)

`- platform: mqtt`
`  name: tesla_rated_range_miles`
`  state_topic: "teslamate/cars/1/rated_battery_range_km"`
`  value_template: "{{ (value | float * 0.621371) |round(2) }}"`
`  unit_of_measurement: 'mi'`
`  icon: mdi:map-marker-path`

`- platform: mqtt`
`  name: tesla_outside_temp_F`
`  state_topic: "teslamate/cars/1/outside_temp"`
`  value_template: "{{ (value | float * 9/5) + 32 |round(2) }}"`
`  unit_of_measurement: F`
`  icon: mdi:thermometer-lines`

**Comment by MrMEScott** [comment]
I have related (EDIT: and stupid) question -- I live in UK but drive in Europe from time to time. When I'm at home the car is set to use imperial units, but when in Europe I like to switch to display km. What happens in Teslamate when it starts getting values in different units?

**Reply by pjnorris** [reply]
looks like values recorded are by default KM, and we have to convert them to MI, so it won't matter.

**Comment by DrJohnM61** [comment]
> I have related question -- I live in UK but drive in Europe from time to time. When I'm at home the car is set to use imperial units, but when in Europe I like to switch to display km. What happens in Teslamate when it starts getting values in different units?

I was just in France the last days (no back in the UK) and the reports in Grafana are shown in miles and look fine (ie, the km traveled is shown correctly/converted to miles).

**Comment by MrMEScott** [comment]
> I have related question -- I live in UK but drive in Europe from time to time. When I'm at home the car is set to use imperial units, but when in Europe I like to switch to display km. What happens in Teslamate when it starts getting values in different units?

I should have just checked the TeslaMate MQTT feed -- everything is in km only. Clearly the conversion to miles is done in the UI - as is implied by there being the units settings on the settings page. And if you look at the JSON behind any of the Grafana displays you can see expressions like `select convert_km(distance, '$length_unit') from drives where id = $drive_id;` - so it's doing the relevant conversions as part of getting the data from the database, where it's all presumably stored as metric values.

**Comment by Dulanic** [comment]
> > I have related question -- I live in UK but drive in Europe from time to time. When I'm at home the car is set to use imperial units, but when in Europe I like to switch to display km. What happens in Teslamate when it starts getting values in different units?
> 
> I should have just checked the TeslaMate MQTT feed -- everything is in km only. Clearly the conversion to miles is done in the UI - as is implied by there being the units settings on the settings page. And if you look at the JSON behind any of the Grafana displays you can see expressions like `select convert_km(distance, '$length_unit') from drives where id = $drive_id;` - so it's doing the relevant conversions as part of getting the data from the database, where it's all presumably stored as metric values.

Tesla seems to prefer km even being a US company. And I don't blame them.
